File: client/src/types/listing.ts
export type ConditionGrade = "DS" | "VNDS" | "USED" | "BEAT";

export type ListingStatus = "draft" | "pending_review" | "accepted" | "rejected";

export type ListingPhoto = {
  id: string;
  url: string;
  sort_order: number;
};

export type SuggestedBidPrice = {
  label: string;
  price: number;
  discount_pct: number;
};

export type Listing = {
  id: string;
  seller_id: string;
  product_id: string | null;
  variant_size: string | null;
  colorway: string | null;
  year_of_release: string | null;
  style_sku: string | null;
  condition_grade: ConditionGrade | null;
  condition_notes: string | null;
  base_price: number | null;
  bid_price: number | null;
  auction_start_at: string | null;
  status: ListingStatus;
  auction_status: AuctionStatus | null;
  winner_id: string | null;
  final_price: number | null;
  sold_at: string | null;
  current_step: number;
  photos: ListingPhoto[];
  suggested_bid_prices: SuggestedBidPrice[];
  created_at: string | null;
  updated_at: string | null;
  submitted_at: string | null;
  reviewed_at: string | null;
};

export type ListingsPage = {
  items: Listing[];
  total: number;
  page: number;
  page_size: number;
};

export type BrowseProductSummary = {
  id: string;
  name: string;
  brand: string | null;
  product_type: string | null;
  price: number | null;
  image_url: string | null;
};

export type BrowseListing = {
  id: string;
  variant_size: string | null;
  condition_grade: ConditionGrade | null;
  bid_price: number | null;
  auction_start_at: string | null;
  auction_status: AuctionStatus | null;
  close_deadline: string | null;
  bid_count: number;
  product: BrowseProductSummary | null;
};

export type BrowseListingsPage = {
  items: BrowseListing[];
  total: number;
  page: number;
  page_size: number;
};

export type AuctionStatus = "scheduled" | "live" | "sold" | "unsold";

export type AuctionScope = "live" | "upcoming" | "sold";

export type Bid = {
  id: string;
  bidder_id: string;
  amount: number;
  created_at: string;
};

export type AuctionSeller = {
  id: string;
  name: string;
  avatar_url: string | null;
};

export type AuctionDetail = {
  id: string;
  product: BrowseProductSummary | null;
  variant_size: string | null;
  condition_grade: ConditionGrade | null;
  condition_notes: string | null;
  photos: ListingPhoto[];
  starting_price: number | null;
  current_price: number;
  min_next_bid: number | null;
  auction_status: AuctionStatus | null;
  auction_start_at: string | null;
  close_deadline: string | null;
  bids: Bid[];
  bid_count: number;
  winner_id: string | null;
  final_price: number | null;
  is_own_listing: boolean;
  view_count: number;
  watch_count: number;
  is_watching: boolean;
  seller: AuctionSeller | null;
};

export type ListingUpdatePatch = Partial<{
  variant_size: string;
  colorway: string;
  year_of_release: string;
  style_sku: string;
  condition_grade: ConditionGrade;
  condition_notes: string;
  bid_price: number;
  auction_start_at: string;
  current_step: number;
}>;

File: server/schemas/listing.py
from typing import Literal

from pydantic import BaseModel, Field

CONDITION_GRADES = ("DS", "VNDS", "USED", "BEAT")
LISTING_STATUSES = ("draft", "pending_review", "accepted", "rejected")

ConditionGrade = Literal["DS", "VNDS", "USED", "BEAT"]


class ListingCreateRequest(BaseModel):
    product_id: str


class ListingUpdateRequest(BaseModel):
    variant_size: str | None = Field(default=None, min_length=1, max_length=100)
    colorway: str | None = Field(default=None, max_length=100)
    year_of_release: str | None = Field(default=None, max_length=20)
    style_sku: str | None = Field(default=None, max_length=100)
    condition_grade: ConditionGrade | None = None
    condition_notes: str | None = Field(default=None, max_length=500)
    bid_price: float | None = Field(default=None, gt=0)
    auction_start_at: str | None = None
    current_step: int | None = Field(default=None, ge=1, le=6)


class ListingReviewRequest(BaseModel):
    action: Literal["accept", "reject"]


class ListingPhotoOut(BaseModel):
    id: str
    url: str
    sort_order: int


class SuggestedBidPrice(BaseModel):
    label: str
    price: float
    discount_pct: float


class ListingOut(BaseModel):
    id: str
    seller_id: str
    product_id: str | None = None
    variant_size: str | None = None
    colorway: str | None = None
    year_of_release: str | None = None
    style_sku: str | None = None
    condition_grade: str | None = None
    condition_notes: str | None = None
    base_price: float | None = None
    bid_price: float | None = None
    auction_start_at: str | None = None
    status: str
    auction_status: str | None = None
    winner_id: str | None = None
    final_price: float | None = None
    sold_at: str | None = None
    current_step: int
    photos: list[ListingPhotoOut] = []
    suggested_bid_prices: list[SuggestedBidPrice] = []
    created_at: str | None = None
    updated_at: str | None = None
    submitted_at: str | None = None
    reviewed_at: str | None = None


class ListingsPage(BaseModel):
    items: list[ListingOut]
    total: int
    page: int
    page_size: int


class ReorderPhotosRequest(BaseModel):
    photo_ids: list[str]


class BrowseProductSummary(BaseModel):
    id: str
    name: str
    brand: str | None = None
    product_type: str | None = None
    price: float | None = None
    image_url: str | None = None


class BrowseListingOut(BaseModel):
    id: str
    variant_size: str | None = None
    condition_grade: str | None = None
    bid_price: float | None = None
    auction_start_at: str | None = None
    auction_status: str | None = None
    close_deadline: str | None = None
    bid_count: int = 0
    product: BrowseProductSummary | None = None


class BrowseListingsPage(BaseModel):
    items: list[BrowseListingOut]
    total: int
    page: int
    page_size: int


class BidOut(BaseModel):
    id: str
    bidder_id: str
    amount: float
    created_at: str


class SellerSummary(BaseModel):
    id: str
    name: str
    avatar_url: str | None = None


class AuctionDetailOut(BaseModel):
    id: str
    product: BrowseProductSummary | None = None
    variant_size: str | None = None
    condition_grade: str | None = None
    condition_notes: str | None = None
    photos: list[ListingPhotoOut] = []
    starting_price: float | None = None
    current_price: float
    min_next_bid: float | None = None
    auction_status: str | None = None
    auction_start_at: str | None = None
    close_deadline: str | None = None
    bids: list[BidOut] = []
    bid_count: int = 0
    winner_id: str | None = None
    final_price: float | None = None
    is_own_listing: bool = False
    view_count: int = 0
    watch_count: int = 0
    is_watching: bool = False
    seller: SellerSummary | None = None


class PlaceBidRequest(BaseModel):
    amount: float = Field(gt=0)
